File: csharp/src/Vista.SDK/Experimental/Transport/TimeSeriesData/DataId.cs
namespace Vista.SDK.Experimental.Transport;

public readonly record struct DataId : IEquatable<DataId>
{
    private readonly int _tag;
    private readonly LocalId? _localId;
    private readonly PMSLocalId? _pmsLocalId;
    private readonly string? _shortId;

    public readonly bool IsLocalId => _tag == 1;
    public readonly bool IsPmsLocalId => _tag == 2;
    public readonly bool IsShortId => _tag == 3;

    public readonly LocalId? LocalId => _tag == 1 ? _localId : null;
    public readonly PMSLocalId? PMSLocalId => _tag == 2 ? _pmsLocalId : null;

    public readonly string? ShortId => _tag == 3 ? _shortId : null;

    public DataId(LocalId value)
    {
        _tag = 1;
        _localId = value;
        _pmsLocalId = null;
        _shortId = null;
    }

    public DataId(PMSLocalId value)
    {
        _tag = 2;
        _localId = null;
        _pmsLocalId = value;
        _shortId = null;
    }

    public DataId(string value)
    {
        _tag = 3;
        _localId = null;
        _pmsLocalId = null;
        _shortId = value;
    }

    public readonly T Match<T>(
        Func<LocalId, T> onLocalId,
        Func<PMSLocalId, T> onPmsLocalId,
        Func<string, T> onShortId
    )
    {
        return _tag switch
        {
            1 => onLocalId(_localId!),
            2 => onPmsLocalId(_pmsLocalId!),
            3 => onShortId(_shortId!),
            _ => throw new InvalidOperationException("Tried to match on invalid DataChannelId"),
        };
    }

    public readonly void Switch(
        Action<LocalId> onLocalId,
        Action<PMSLocalId> onPmsLocalId,
        Action<string> onShortId
    )
    {
        if (_tag == 1)
            onLocalId(_localId!);
        else if (_tag == 2)
            onPmsLocalId(_pmsLocalId!);
        else if (_tag == 3)
            onShortId(_shortId!);
        else
            throw new InvalidOperationException("Tried to switch on invalid DataChannelId");
    }

    public override readonly string ToString() =>
        _tag switch
        {
            1 => _localId!.ToString(),
            2 => _pmsLocalId!.ToString(),
            3 => _shortId!,
            _ => throw new Exception("Invalid state exception"),
        };

    public static DataId Parse(string value)
    {
        if (value is null)
            throw new ArgumentNullException(nameof(value));
        if (LocalIdBuilder.TryParse(value, out var localIdBuilder))
            return new DataId(localIdBuilder.Build());
        else if (PMSLocalIdBuilder.TryParse(value, out var pmsLocalIdBuilder))
            return new DataId(pmsLocalIdBuilder.Build());
        else
            return new DataId(value);
    }

    public override int GetHashCode()
    {
        var hash = default(HashCode);

        Switch(
            l => hash.Add(l.GetHashCode()),
            pl => hash.Add(pl.GetHashCode()),
            s => hash.Add(s.GetHashCode())
        );

        return hash.ToHashCode();
    }

    public bool Equals(DataId other) =>
        _tag switch
        {
            1 => other.IsLocalId && other.LocalId == _localId,
            2 => other.IsPmsLocalId && other.PMSLocalId == _pmsLocalId,
            3 => other.IsShortId && other.ShortId == _shortId,
            _ => throw new Exception("Invalid state exception"),
        };
}

File: js/lib/experimental/transport/domain/time-series-data/DataId.ts
import { Codebooks } from "../../../../Codebooks";
import { Gmod } from "../../../../Gmod";
import { LocalId } from "../../../../LocalId";
import { LocalIdBuilder } from "../../../../LocalId.Builder";
import { Locations } from "../../../../Location";
import { LocalIdParsingErrorBuilder } from "../../../../internal/LocalIdParsingErrorBuilder";
import { isNullOrWhiteSpace } from "../../../../util/util";
import { PMSLocalId } from "../../../PMSLocalId";
import { PMSLocalIdBuilder } from "../../../PMSLocalId.Builder";

export class DataId {
    private readonly _tag: number;
    private readonly _localId?: LocalId;
    private readonly _pmsLocalId?: PMSLocalId;
    private readonly _shortId?: string;

    public constructor(value: LocalId | PMSLocalId | string) {
        if (value instanceof LocalId) {
            this._tag = 1;
            this._localId = value;
        } else if (value instanceof PMSLocalId) {
            this._tag = 2;
            this._pmsLocalId = value;
        } else {
            this._tag = 3;
            this._shortId = value;
        }
    }

    public get isLocalId(): boolean {
        return this._tag == 1;
    }

    public get isPMSLocalId(): boolean {
        return this._tag == 2;
    }

    public get isShortId(): boolean {
        return this._tag == 3;
    }

    public get localId() {
        return this._tag === 1 ? this._localId : undefined;
    }

    public get pmsLocalId() {
        return this._tag === 2 ? this._pmsLocalId : undefined;
    }

    public get shortId() {
        return this._tag === 3 ? this._shortId : undefined;
    }

    public match<T>(
        onLocalId: (localId: LocalId) => T,
        onPMSLocalId: (pmsLocalId: PMSLocalId) => T,
        onShortId: (shortId: string) => T
    ): T {
        switch (this._tag) {
            case 1:
                return onLocalId(this._localId!);
            case 2:
                return onPMSLocalId(this._pmsLocalId!);
            case 3:
                return onShortId(this._shortId!);
            default:
                throw new Error("Tried to match on invalid DataChannelId");
        }
    }

    public switch(
        onLocalId: (localId: LocalId) => void,
        onPMSLocalId: (pmsLocalId: PMSLocalId) => void,
        onShortId: (shortId: string) => void
    ): void {
        switch (this._tag) {
            case 1:
                onLocalId(this._localId!);
                break;
            case 2:
                onPMSLocalId(this._pmsLocalId!);
                break;
            case 3:
                onShortId(this._shortId!);
                break;
            default:
                throw new Error("Tried to switch on invalid DataChannelId");
        }
    }

    public toString(): string {
        switch (this._tag) {
            case 1:
                return this._localId!.toString();
            case 2:
                return this._pmsLocalId!.toString();
            case 3:
                return this._shortId!;
            default:
                throw new Error("Invalid DataChannelId");
        }
    }

    public static parse(
        value: string | undefined,
        gmod: Gmod,
        codebooks: Codebooks,
        locations: Locations,
        errorBuilder?: LocalIdParsingErrorBuilder
    ): DataId {
        if (isNullOrWhiteSpace(value))
            throw new Error(`${this.parse.name}: value is null`);

        let localId: LocalIdBuilder | undefined = undefined;
        let pmsLocalId: PMSLocalIdBuilder | undefined = undefined;

        if (
            !!(localId = LocalIdBuilder.tryParse(
                value,
                gmod,
                codebooks,
                locations,
                errorBuilder
            ))
        ) {
            return new DataId(localId.build());
        } else if (
            !!(pmsLocalId = PMSLocalIdBuilder.tryParse(
                value,
                gmod,
                codebooks,
                locations,
                errorBuilder
            ))
        ) {
            return new DataId(pmsLocalId.build());
        } else {
            return new DataId(value);
        }
    }

    public static async parseAsync(
        value: string | undefined,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        if (isNullOrWhiteSpace(value))
            throw new Error(`${this.parse.name}: value is null`);

        let localId: LocalIdBuilder | undefined = undefined;
        let pmsLocalId: PMSLocalIdBuilder | undefined = undefined;

        if (
            (localId = await LocalIdBuilder.tryParseAsync(value, errorBuilder))
        ) {
            return new DataId(localId.build());
        } else if (
            (pmsLocalId = await PMSLocalIdBuilder.tryParseAsync(
                value,
                errorBuilder
            ))
        ) {
            return new DataId(pmsLocalId.build());
        } else {
            return new DataId(value);
        }
    }
}
